softdevteam/eco
# Eco plugin API for external tools

This PR provides a plugin API for Eco. The aim is to allow users to annotate
their code with information from external tools.

Advantages of this approach are:
- External tools no longer piggyback on the export / run code, which leaves
  that code cleaner and obviates the need for the `mock` objects which
  were a work-around to make the console work with the CPython profiler.
- Plugins are stored in `~/.eco/` which means that different users can have
  their own tools set up in different ways, according to their own tastes.
- File paths to external tools, which are likely to differ between users,
  are now stored externally in the plugins, and not hard-coded into Eco.
  The exception to this is the **Project -> Run** command which this PR
  does not touch.
- Different users can configure the same tools to produce different
  visualisations (by using _Hints_ differently).
- The **Tools** menu is generated at run-time, according to which plugins
  are loaded.
- Plugins are just ordinary Python code, which should conform to the
  documented API.
- External tools run on a timer and in a separate thread to the main GUI.
- The throbber spins when the user runs a program, not when the external
  tools are running.

In order to test this PR you will need to:
- Fetch and checkout the `eco-plugins` branch
- Create a directory called `~/.eco/`
- Copy the files in `plugins/` into your `~/.eco/` directory
- Open Eco. Check that the **Tools** menu corresponds to your plugins.
  There should be sub-menus for each language which has a plugin.
- The Python plugins should work for everyone and only depend on stdlib.
- Open or create a Python file. You should see tooltips and a heatmap.
- Try turning the Python plugins on and off.

Some subtleties to look out for:
- When more than one document is open in Eco, do the plugins behave
  correctly when you change tabs?
- By design, if a plugin fails after it has been loaded, nothing happens.
  A plugin is a proxy for an external tool, so if one errors, it should not
  interrupt the user. It might be worth making a plugin fail somehow to
  check this.
- Will the right plugins be loaded for the right languages? At the moment
  the `self.language` field of each plugin (which correspond to sub-menus
  under the **Tools** menu) must exactly match the language of the currently
  open document in order to run.
- Does the documentation make sense?

Some things this PR does not do, but could in the future:
- Save the state of the plugins per-document. I.e. whenever you open a
  document the Eco tab starts with all plugins activated, regardless of how
  you used plugins the last time you opened the document.
- Expose many internals to the user.
- Allow you to check different plugins for different tabs.
- If a node has more than one _Footnote_ or _Tooltip_ annotation, the
  text from the annotations is concatenated. However, only one sort of
  annotation can make use of the heatmap overlay. To date, we have never had
  two plugins generating heatmap data, but if this were to happen the user
  would have to deactivate one of their plugins. This may be the right level
  flexibilty, but the alternative is to add some extra logic to Eco so that
  it chooses which data to display for the user.

Fixes #35
